File: leetcode/C++/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* t;
    ListNode* swapPairs(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode *prev=head->next;
        t=head->next->next;
        prev->next=head;
        head->next=swapPairs(t);
        return prev;
    }
};

File: leetcode/Python/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def __init__(self):
        t=None
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        prev=head.next
        self.t, prev.next = prev.next, head
        head.next=self.swapPairs(self.t)
        return prev